File: apps/demo-be-rsax/tests/integration/steps/token_lifecycle_steps.rs
use cucumber::{given, then, when};

use crate::world::{get_req, json_req, AppWorld};

#[when("alice sends POST /api/v1/auth/refresh with her refresh token")]
async fn refresh_with_alice_token(world: &mut AppWorld) {
    let token = world.refresh_token.clone().unwrap_or_default();
    let body = format!(r#"{{"refresh_token": "{token}"}}"#);
    let req = json_req("POST", "/api/v1/auth/refresh", &body, None);
    world.send(req).await.unwrap();
    if world.last_status == 200 {
        world.auth_token = world
            .last_body
            .get("access_token")
            .and_then(|v| v.as_str())
            .map(String::from);
        world.refresh_token = world
            .last_body
            .get("refresh_token")
            .and_then(|v| v.as_str())
            .map(String::from);
    }
}

#[given("alice's refresh token has expired")]
async fn alice_refresh_expired(world: &mut AppWorld) {
    // Use an obviously expired/invalid token
    world.refresh_token = Some(
        "eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJzdWIiOiJ0ZXN0IiwiZXhwIjoxfQ.invalid".to_string(),
    );
}

#[given("alice has used her refresh token to get a new token pair")]
async fn alice_used_refresh_token(world: &mut AppWorld) {
    let original = world.refresh_token.clone().unwrap_or_default();
    world.original_refresh_token = Some(original.clone());

    let body = format!(r#"{{"refresh_token": "{original}"}}"#);
    let req = json_req("POST", "/api/v1/auth/refresh", &body, None);
    world.send(req).await.unwrap();
    if world.last_status == 200 {
        world.auth_token = world
            .last_body
            .get("access_token")
            .and_then(|v| v.as_str())
            .map(String::from);
        world.refresh_token = world
            .last_body
            .get("refresh_token")
            .and_then(|v| v.as_str())
            .map(String::from);
    }
}

#[when("alice sends POST /api/v1/auth/refresh with her original refresh token")]
async fn refresh_with_original_token(world: &mut AppWorld) {
    let token = world.original_refresh_token.clone().unwrap_or_default();
    let body = format!(r#"{{"refresh_token": "{token}"}}"#);
    let req = json_req("POST", "/api/v1/auth/refresh", &body, None);
    world.send(req).await.unwrap();
}

#[given(expr = "the user {string} has been deactivated")]
async fn user_deactivated_step(world: &mut AppWorld, _username: String) {
    // Deactivate alice using admin or using alice's own token
    let token = world.auth_token.clone().unwrap_or_default();
    let req = json_req(
        "POST",
        "/api/v1/users/me/deactivate",
        "{}",
        Some(&format!("Bearer {token}")),
    );
    world.send(req).await.unwrap();
}

#[when("alice sends POST /api/v1/auth/logout with her access token")]
async fn logout_alice(world: &mut AppWorld) {
    let token = world.auth_token.clone().unwrap_or_default();
    let req = json_req(
        "POST",
        "/api/v1/auth/logout",
        "{}",
        Some(&format!("Bearer {token}")),
    );
    world.send(req).await.unwrap();
}

#[when("alice sends POST /api/v1/auth/logout-all with her access token")]
async fn logout_all_alice(world: &mut AppWorld) {
    let token = world.auth_token.clone().unwrap_or_default();
    let req = json_req(
        "POST",
        "/api/v1/auth/logout-all",
        "{}",
        Some(&format!("Bearer {token}")),
    );
    world.send(req).await.unwrap();
}

#[then("alice's access token should be invalidated")]
async fn check_token_invalidated(world: &mut AppWorld) {
    let token = world.auth_token.clone().unwrap_or_default();
    let req = get_req("/api/v1/users/me", Some(&format!("Bearer {token}")));
    world.send(req).await.unwrap();
    assert_eq!(
        world.last_status, 401,
        "Expected 401 after logout, got {}",
        world.last_status
    );
}

#[given("alice has already logged out once")]
async fn alice_already_logged_out(world: &mut AppWorld) {
    let token = world.auth_token.clone().unwrap_or_default();
    let req = json_req(
        "POST",
        "/api/v1/auth/logout",
        "{}",
        Some(&format!("Bearer {token}")),
    );
    world.send(req).await.unwrap();
}

#[then("the response body should contain an error message about token expiration")]
async fn error_token_expired(world: &mut AppWorld) {
    let msg = world
        .last_body
        .get("message")
        .and_then(|v| v.as_str())
        .unwrap_or("");
    assert!(
        !msg.is_empty(),
        "Expected token expiration error, got: {}",
        world.last_body
    );
}

#[then("the response body should contain an error message about invalid token")]
async fn error_invalid_token(world: &mut AppWorld) {
    let msg = world
        .last_body
        .get("message")
        .and_then(|v| v.as_str())
        .unwrap_or("");
    assert!(
        !msg.is_empty(),
        "Expected invalid token error, got: {}",
        world.last_body
    );
}

File: apps/a-demo-be-elixir-phoenix/test/unit/steps/token_lifecycle_steps.exs
defmodule AADemoBeExphWeb.Unit.TokenLifecycleSteps do
  use Cabbage.Feature, async: false, file: "authentication/token-lifecycle.feature"

  use AADemoBeExphWeb.ConnCase

  alias ADemoBeExph.Integration.Helpers

  @moduletag :unit

  defp accounts, do: Application.get_env(:a_demo_be_exph, :accounts_module)
  defp token_ctx, do: Application.get_env(:a_demo_be_exph, :token_module)

  defgiven ~r/^the API is running$/, _vars, state do
    {:ok, state}
  end

  defgiven ~r/^a user "(?<username>[^"]+)" is registered with password "(?<password>[^"]+)"$/,
           %{username: username, password: password},
           state do
    email = "#{username}@example.com"
    user = Helpers.register_user!(username, email, password)
    {:ok, Map.merge(state, %{alice: user, alice_password: password})}
  end

  defgiven ~r/^"(?<username>[^"]+)" has logged in and stored the access token and refresh token$/,
           %{username: _username},
           %{alice: user} = state do
    {access_token, refresh_token} = Helpers.login_user!(user)
    {:ok, Map.merge(state, %{access_token: access_token, refresh_token: refresh_token})}
  end

  defgiven ~r/^alice's refresh token has expired$/,
           _vars,
           %{refresh_token: refresh_token} = state do
    token_ctx().expire_refresh_token!(refresh_token)
    {:ok, state}
  end

  defgiven ~r/^alice has used her refresh token to get a new token pair$/,
           _vars,
           %{alice: user, refresh_token: refresh_token} = state do
    token_ctx().consume_refresh_token(refresh_token)
    {:ok, new_refresh_token} = token_ctx().create_refresh_token(user.id)
    {:ok, Map.put(state, :new_refresh_token, new_refresh_token)}
  end

  defgiven ~r/^the user "(?<username>[^"]+)" has been deactivated$/,
           %{username: _username},
           %{alice: user} = state do
    accounts().deactivate_user(user)
    {:ok, state}
  end

  defgiven ~r/^alice has already logged out once$/,
           _vars,
           %{alice: _user, access_token: access_token} = state do
    build_conn()
    |> put_req_header("authorization", "Bearer #{access_token}")
    |> post("/api/v1/auth/logout", %{})

    {:ok, state}
  end

  defwhen ~r/^alice sends POST \/api\/v1\/auth\/refresh with her refresh token$/,
          _vars,
          %{refresh_token: refresh_token} = state do
    body = Jason.encode!(%{"refreshToken" => refresh_token})

    conn =
      build_conn()
      |> put_req_header("content-type", "application/json")
      |> post("/api/v1/auth/refresh", body)

    {:ok, Map.put(state, :conn, conn)}
  end

  defwhen ~r/^alice sends POST \/api\/v1\/auth\/refresh with her original refresh token$/,
          _vars,
          %{refresh_token: refresh_token} = state do
    body = Jason.encode!(%{"refreshToken" => refresh_token})

    conn =
      build_conn()
      |> put_req_header("content-type", "application/json")
      |> post("/api/v1/auth/refresh", body)

    {:ok, Map.put(state, :conn, conn)}
  end

  defwhen ~r/^alice sends POST \/api\/v1\/auth\/logout with her access token$/,
          _vars,
          %{access_token: access_token} = state do
    conn =
      build_conn()
      |> put_req_header("authorization", "Bearer #{access_token}")
      |> post("/api/v1/auth/logout", %{})

    {:ok, Map.put(state, :conn, conn)}
  end

  defwhen ~r/^alice sends POST \/api\/v1\/auth\/logout-all with her access token$/,
          _vars,
          %{access_token: access_token} = state do
    conn =
      build_conn()
      |> put_req_header("authorization", "Bearer #{access_token}")
      |> post("/api/v1/auth/logout-all", %{})

    {:ok, Map.put(state, :conn, conn)}
  end

  defthen ~r/^the response status code should be (?<code>\d+)$/,
          %{code: code},
          %{conn: conn} = state do
    assert conn.status == String.to_integer(code)
    {:ok, state}
  end

  defthen ~r/^the response body should contain a non-null "(?<field>[^"]+)" field$/,
          %{field: field},
          %{conn: conn} = state do
    body = Jason.decode!(conn.resp_body)
    assert Map.has_key?(body, field)
    assert body[field] != nil
    {:ok, state}
  end

  defthen ~r/^the response body should contain an error message about token expiration$/,
          _vars,
          %{conn: conn} = state do
    body = Jason.decode!(conn.resp_body)
    assert body["message"] =~ ~r/[Ee]xpir/i
    {:ok, state}
  end

  defthen ~r/^the response body should contain an error message about invalid token$/,
          _vars,
          %{conn: conn} = state do
    body = Jason.decode!(conn.resp_body)
    assert body["message"] =~ ~r/[Ii]nvalid/i
    {:ok, state}
  end

  defthen ~r/^the response body should contain an error message about account deactivation$/,
          _vars,
          %{conn: conn} = state do
    body = Jason.decode!(conn.resp_body)
    assert body["message"] =~ ~r/[Dd]eactivat|[Ii]nactive|[Dd]isabled/i
    {:ok, state}
  end

  defthen ~r/^alice's access token should be invalidated$/,
          _vars,
          %{access_token: access_token} = state do
    conn =
      build_conn()
      |> put_req_header("authorization", "Bearer #{access_token}")
      |> get("/api/v1/users/me")

    assert conn.status == 401
    {:ok, state}
  end
end
